src/jit/emitxarch.cpp

```diff
@@ -2901,174 +2901,283 @@ void emitter::emitInsStoreLcl(instruction ins, emitAttr attr, GenTreeLclVarCommo
 
 regNumber emitter::emitInsBinary(instruction ins, emitAttr attr, GenTree* dst, GenTree* src)
 {
-    // dst can only be a reg or modrm
-    assert(!dst->isContained() || dst->isUsedFromMemory() || instrIs3opImul(ins)); // dst on these isn't really the dst
+    // We can only have one memory operand and only src can be a constant operand
+    GenTree* memOp = nullptr;
+    GenTree* cnsOp = nullptr;
+    GenTree* otrOp = nullptr;
 
-#ifdef DEBUG
-    // src can be anything but both src and dst cannot be addr modes
-    // or at least cannot be contained addr modes
-    if (dst->isUsedFromMemory())
+    if (dst->isContained() || (dst->isLclField() && (dst->gtRegNum == REG_NA)) || dst->isUsedFromSpillTemp())
     {
+        // dst can only be a reg or modrm
+        assert(dst->isUsedFromMemory() || (dst->gtRegNum == REG_NA) ||
+               instrIs3opImul(ins)); // dst on 3opImul isn't really the dst
         assert(!src->isUsedFromMemory());
-    }
-
-    if (src->isUsedFromMemory())
-    {
-        assert(!dst->isUsedFromMemory());
-    }
-#endif
 
-    // find which operand is a memory op (if any)
-    // and what its base is
-    GenTreeIndir* mem     = nullptr;
-    GenTree*      memBase = nullptr;
+        memOp = dst;
 
-    if (dst->isContainedIndir())
-    {
-        mem = dst->AsIndir();
+        if (src->isContained())
+        {
+            assert(src->IsCnsIntOrI());
+            cnsOp = src;
+        }
+        else
+        {
+            otrOp = src;
+        }
     }
-    else if (src->isContainedIndir())
+    else if (src->isContained() || src->isUsedFromSpillTemp())
     {
-        mem = src->AsIndir();
-    }
+        assert(!dst->isUsedFromMemory());
+        otrOp = dst;
 
-    if (mem)
-    {
-        memBase = mem->gtOp1;
+        if (src->IsCnsIntOrI() || src->IsCnsFltOrDbl())
+        {
+            cnsOp = src;
```

Comment:
I originally had a `assert(!src->isUsedFromMemory() || src->IsCnsFltOrDbl());` right above this. However, it appears that on x86 Checked builds, we will sometimes spill a GT_CNS_INT to the stack.

See the failure [here](https://ci.dot.net/job/dotnet_coreclr/job/master/job/x86_checked_windows_nt_prtest/14191/) which reported:
```
Assert failure(PID 4672 [0x00001240], Thread: 4644 [0x1224]): Assertion failed '!src->isUsedFromMemory() || src->IsCnsFltOrDbl()' in 'System.Diagnostics.Tracing.EventProvider:WriteEvent(byref,int,int,int,ref):bool:this' (IL size 1225)
```

Reply:
I was not seeing the same failure in x86 Debug or Release.

Reply:
Looks like x86 Checked is still failing due to the stack spill. Will investigate sometime tomorrow.

Reply:
`isUsedFromSpillTemp` needed to take precedence over everything else (including `IsCns` and `IsIndir`).